beacon-chain/sync/rpc_data_column_sidecars_by_root.go

```diff
@@ -0,0 +1,182 @@
+
+// dataColumnSidecarByRootRPCHandler handles the data column sidecars by root RPC request.
+// https://github.com/ethereum/consensus-specs/blob/dev/specs/fulu/p2p-interface.md#datacolumnsidecarsbyroot-v1
+func (s *Service) dataColumnSidecarByRootRPCHandler(ctx context.Context, msg interface{}, stream libp2pcore.Stream) error {
+	ctx, span := trace.StartSpan(ctx, "sync.dataColumnSidecarByRootRPCHandler")
+	defer span.End()
+
+	batchSize := flags.Get().DataColumnBatchLimit
+	numberOfColumns := params.BeaconConfig().NumberOfColumns
+
+	// Check if the message type is the one expected.
+	ref, ok := msg.(*types.DataColumnsByRootIdentifiers)
+	if !ok {
+		return notDataColumnsByRootIdentifiersError
+	}
+
+	requestedColumnIdents := *ref
+	remotePeerId := stream.Conn().RemotePeer()
+
+	ctx, cancel := context.WithTimeout(ctx, ttfbTimeout)
+	defer cancel()
+
+	SetRPCStreamDeadlines(stream)
+
+	// Penalize peers that send invalid requests.
+	if err := validateDataColumnsByRootRequest(requestedColumnIdents); err != nil {
+		s.cfg.p2p.Peers().Scorers().BadResponsesScorer().Increment(remotePeerId)
+		s.writeErrorResponseToStream(responseCodeInvalidRequest, err.Error(), stream)
+		return errors.Wrap(err, "validate data columns by root request")
+	}
+
+	requestedColumnsByRoot := make(map[[fieldparams.RootLength]byte][]uint64)
+	for _, columnIdent := range requestedColumnIdents {
+		var root [fieldparams.RootLength]byte
+		copy(root[:], columnIdent.BlockRoot)
+		requestedColumnsByRoot[root] = append(requestedColumnsByRoot[root], columnIdent.Columns...)
+	}
+
+	// Sort by column index for each root.
+	for _, columns := range requestedColumnsByRoot {
+		slices.Sort(columns)
+	}
+
+	// Format nice logs.
+	requestedColumnsByRootLog := make(map[string]interface{})
+	for root, columns := range requestedColumnsByRoot {
+		rootStr := fmt.Sprintf("%#x", root)
+		requestedColumnsByRootLog[rootStr] = "all"
+		if uint64(len(columns)) != numberOfColumns {
+			requestedColumnsByRootLog[rootStr] = columns
+		}
+	}
+
+	// Compute the oldest slot we'll allow a peer to request, based on the current slot.
+	minReqSlot, err := dataColumnsRPCMinValidSlot(s.cfg.clock.CurrentSlot())
+	if err != nil {
+		return errors.Wrapf(err, "data columns RPC min valid slot")
+	}
+
+	log := log.WithFields(logrus.Fields{
+		"peer":    remotePeerId,
+		"columns": requestedColumnsByRootLog,
+	})
+
+	defer closeStream(stream, log)
+
+	var ticker *time.Ticker
+	if len(requestedColumnIdents) > batchSize {
+		ticker = time.NewTicker(time.Second)
+	}
+
+	log.Debug("Serving data column sidecar by root request")
+
+	count := 0
+	for root, columns := range requestedColumnsByRoot {
+		if err := ctx.Err(); err != nil {
+			closeStream(stream, log)
+			return errors.Wrap(err, "context error")
+		}
+
+		// Throttle request processing to no more than batchSize/sec.
+		for range columns {
+			if ticker != nil && count != 0 && count%batchSize == 0 {
```

Comment:
does count ever need to be reset? ( maybe not if the number doesn't get too high)

Reply:
but also do you think it'd be benefitial to use the rate limiter in the time package instead? golang.org/x/time/rate

Reply:
maybe not just saw s.rateLimiter.add

Reply:
> does count ever need to be reset? ( maybe not if the number doesn't get too high)

`count` is initialized to `0` at each `dataColumnSidecarByRootRPCHandler` call.

It counts the total number of columns.